README.md
# Braket-template

[![GitHub version](https://badge.fury.io/gh/danlevan%2Fbracket-template.svg)](http://badge.fury.io/gh/danlevan%2Fbracket-template)
[![npm version](https://badge.fury.io/js/bracket-template.svg)](http://badge.fury.io/js/bracket-template)

> Minimal (über fast) Javascript engine compatible with node.js and browsers.

## TL;DR

- Javascript templating lib for node and modern browsers 
- Ultra fast, ultra small, ultra fast to learn 
- Uses `[[ ]]` by default (customizable), so plays well with `{{ }}` (Angular, Ember...)
- Define and call your own block definition (functions) for reuse
- Node: Parent layout and include files supported

## Getting started

`$ npm i bracket-template`

Or, fork on [github](https://github.com/danlevan/bracket-template)

## Express

If you want to use bracket with express, it's recommended to use [consolidate.js](https://www.npmjs.com/package/consolidate) as it makes is easy to change engine and implements a cache.

[example](examples/node/consolidate.js)

## Node (API)

If you want to use bracket in your node project for building your email templates for example, you can use it directly. (For Express, it's recommended to use [consolidate.js](https://www.npmjs.com/package/consolidate))

[example](examples/node/simple.js)

## Browser

Bracket works great with modern browsers. The tests are run on the latest chrome browser (more browser tests to come).

[example](examples)

## Language definition

### Javascript execution

* Anything between `[[ ... ]]` is executed as javascript.
* Anything bewteen `[[= ... ]]` is interpolated (i.e. print string values)

For example

```
[[ var test = 'wor'; ]]
Hello [[= test + '..ld' ]]
```

→ **Result** `Hello wor..ld`

### Block definition

* Block definition `[[## someFunc(someArg...) ... #]]`
* Call blocks with `[[# someFunc(...) ]]`

For example

```html
<!-- Call block -->
Hello [[# block1('world') ]]

<!-- Block definition -->
[[## block1(arg1)
  from block1 (with '[[= arg1 ]]')
#]]
```

→ **Result** `Hello from block 1 (with 'world')`

### Helper methods

Helper methods can be passed during the compilation of the template. This is especially useful for passing a translation method for example.

Example

```javascript
var template = bracket.compile(
  "Hello world: [[# translate('Hello world') ]]",
  {
    helpers: {
      translate: function(text) {
        return 'Bonjour le monde';
      }
    }
  }
);
template();
```

→ **Result** `Hello world: Bonjour le monde`

## Extras for node

When used in node, you can define a master layout and include partial files.

### Layout 

Layout declarations are done in [yaml](http://yaml.org) at the beginning of the template file.

**index.dot**

```html
---
master: master.dot
title: Index page
---

<!-- Define blocks used in master.dot -->

[[## body()
  Hello from index.dot
#]]

[[## body2()
  Hello from index.dot again
#]]
```

**master.dot**

```html
<!doctype html>
<html lang="en">
  <head>
    <title>[[= layout.title || 'default title' ]]</title>
  </head>
  <body>
    Hello from master.dot <br />
    [[# body() ]] <br />
    [[# body2() ]]
  </body>
</html>
```

→ **Result**

```html
<!doctype html>
<html lang="en">
  <head>
    <title>Index page</title>
  </head>
  <body>
    Hello from master.dot <br />
    Hello from index.dot <br />
    Hello from index.dot again
  </body>
</html>
```

**Pro tip**
- Custom variables can be defined in the yaml (like title in the example) and used in the master.
- Default values in this format `[[= layout.title || 'default title' ]]`

### Partials

You can include another file in your template file.

**index.dot**

```html
<!doctype html>
<html lang="en">
  <head>
    <title>Example</title>
  </head>
  <body>
    [[## partial('header.dot', { current: 'contact' }) ]]
  </body>
</html>
```

**header.dot**

```html
<nav>
  <ul>
    <li>Home</li>
    <li>Contact</li>
  </ul>
</nav>
```

→ **Result**

```html
<!doctype html>
<html lang="en">
  <head>
    <title>Example</title>
  </head>
  <body>
    <nav>
      <ul>
        <li>Home</li>
        <li>Contact</li>
      </ul>
    </nav>
  </body>
</html>
```

## Customization

You can customize all the settings and regex used by bracket. 

**Default settings**

```js
const settings = {
  // Extract anything inside [[ ]] to be evaluated as js
  evaluate: /\[\[([\s\S]+?(\}?)+)]]/g,

  // Extract anything in the form [[= model ]]
  interpolate: /\[\[=([\s\S]+?)]]/g,

  // Extract any block call [[# block1('arg') ]]
  block: /\[\[#\s*([\w]+)\(([\s\S]*?)\)\s*]]/g,

  // Extract any block definition [[## block1(arg) #]]
  blockDef: /\[\[##\s*([\w]+)\(([\s\w,]*)\)\s*[\n]([\s\S]*?)\n\s*#]]/g,

  // extract the argument values from a function call
  // e.g. { test1: '123', test2: 456, test3: true }, 'aaa', true, {}, ''
  argValues: /\s*({[\s\S]*?}|[^,]+)/g,

  // The params to pass to the template function
  // For multiple params, comma delimited e.g. 'model,model2,model3...'
  varname: 'model',
};
```

**Pro tip** If you want to pass more than one model, change the varname to 'model, model2' and your template function will accept 2 models.


CHANGELOG.md
# Versions

## v1.1.0 (december 2016)

* Support helpers
* Node version (support layout and partials)

## v1.0.0 (november 2016)

* Initial version


src/bracket.js
/*
 * Bracket template library
 *
 */

/* global window VERSION */

const settings = {
  // Extract anything inside [[ ]] to be evaluated as js
  evaluate: /\[\[([\s\S]+?(\}?)+)]]/g,

  // Extract anything in the form [[= model ]]
  interpolate: /\[\[=([\s\S]+?)]]/g,

  // Extract any block call [[# block1('arg') ]]
  block: /\[\[#\s*([\w]+)\(([\s\S]*?)\)\s*]]/g,

  // Extract any block definition [[## block1(arg) #]]
  blockDef: /\[\[##\s*([\w]+)\(([\s\w,]*)\)\s*[\n]([\s\S]*?)\n\s*#]]/g,

  // extract the argument values from a function call
  // e.g. { test1: '123', test2: 456, test3: true }, 'aaa', true, {}, ''
  argValues: /\s*({[\s\S]*?}|[^,]+)/g, // TODO: bug with , and }

  // The params to pass to the template function
  // For multiple params, comma delimited e.g. 'model,model2,model3...'
  varname: 'model',

  // helper functions
  helpers: {},
};

function handleBlockCall(c, blocks) {
  return (m, name, argStr) => {
    // no block definition
    if (!c.helpers[name] && !blocks[name]) {
      return '';
    }

    // split arg string
    const args = [];
    argStr.replace(c.argValues, (m2, val) => {
      args.push(val);
    });

    // call helpers
    if (c.helpers[name]) {
      const val = c.helpers[name](...args.map(a => Function(`return ${a};`)())); // eslint-disable-line
      return `';out+=${JSON.stringify(val)};out+='`;
    }

    // call block def

    // maps arg name and value
    const lookup = blocks[name].args.reduce((res, k, i) => {
      const hash = res;
      hash[k] = args.length <= i ? undefined : args[i].replace(/\\'/g, '\''); // fix the replacement of ' to \'
      return hash;
    }, {});

    const blockStr = blocks[name].body
      .replace(/'/g, '\\\'') // allow the use of ' in the body
      // replace block def with arg values
      .replace(c.interpolate, (m2, codeVal) => {
        const code = codeVal.trim();

        // support obj.arg
        const key = code.split('.')[0];

        // key not found then leave it as is
        if (!(key in lookup)) {
          return m2;
        }

        // Generate the value by making a scoped function
        // e.g. function() { var arg = { test1: '123' }; return arg.test1; }
        const valStr = `var ${key}=${lookup[key]};return ${code};`;
        const val = Function(valStr)(); // eslint-disable-line

        return `';out+=${JSON.stringify(val)};out+='`;
      })
      // recursive handle block call
      .replace(c.block, handleBlockCall(c, blocks));

    return blockStr;
  };
}

/**
 * Compile creates a template function
 * @param {string} tmpl - The html template to compile
 * @param {object} conf - The configuration to override
 * @return {function} The template function
 */
function compile(tmpl, conf) {
  let str = tmpl || '';
  const c = Object.assign({}, settings, conf);
  const blocks = {};

  str = str
    // handle block def
    .replace(c.blockDef, (m, name, argStr, body) => {
      blocks[name] = {
        args: argStr.split(',').map(a => a.trim()),
        body,
      };
      return '';
    })

    // handle block call
    .replace(c.block, handleBlockCall(c, blocks))

    // convert models
    .replace(c.interpolate, (m, code) => `';out+=${code.trim()};out+='`)

    // raw js
    .replace(c.evaluate, (m, code) => `';${code.trim()}out+='`);

  // build the template function body
  str = `var out='${str}';return out;`
    // remove the newlines or '' will break
    .replace(/\r/g, '\\r')
    .replace(/\n/g, '\\n')
    .replace(/\t/g, '\\t')
    .replace(/(\s|;|\}|^|\{)out\+='';/g, '$1')
    .replace(/\+''/g, '');

  return new Function(c.varname, str); // eslint-disable-line
}

// The object to export
const res = {
  version: typeof VERSION === 'undefined' ? 'test' : VERSION, // read from package.json
  settings,
  compile,
};

// browser: add to window.bracket
if (typeof window !== 'undefined') {
  window.bracket = res;
}

export default res;
